### PR TITLE
DSI uses ENV

### PR DESCRIPTION
**Pull request by d-a-v-e**
## Changes in this PR

- consolidate DSI linking building logic into a dedicated class
- return a URI that defaults to the production environment
- add `DSI_ENV` to control the environment and link prefix
- remove `DFE_SIGN_IN_ISSUER` and `DFE_SIGN_IN_API_ENDPOINT`

## TODO

- delete old environment variables from the GH settings

### CODE REVIEW

**Comment by peterdavidhamilton** [comment]
[DSI env for use in link logic](https://trello.com/c/KTT97ahO/780-dsi-env-for-use-in-link-logic)

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -11,6 +11,6 @@ def banner_message
   end
 
   def dsi_url(args = {})
```
Comment (peterdavidhamilton) [inline app/helpers/application_helper.rb:13]: Can we slurp and splat on this?

File: lib/dsi/client.rb [hunk 2]
```diff
@@ -148,7 +148,7 @@ def get_users
 
     # @return [URI::HTTPS]
```
Comment (peterdavidhamilton) [inline lib/dsi/client.rb:149]: Can you document the `@param`

File: spec/lib/self-serve/dsi/uri_spec.rb [hunk 3]
```diff
@@ -0,0 +1,147 @@
+require "dsi/uri"
+
+RSpec.describe Dsi::Uri do
+  subject(:service) { described_class.new(**args) }
+
+  # TODO: replace meta-programmed testing to explicit context blocks
```
Comment (peterdavidhamilton) [inline spec/lib/self-serve/dsi/uri_spec.rb:6]: This is even more complex than before.

File: spec/lib/self-serve/dsi/uri_spec.rb [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        context "with no path" do
+          it { expect(service.call).to eql URI("https://test-services.signin.education.gov.uk/") }
+        end
+      end
+    end
+
+    context "when no DSI_ENV has been set" do
+      # rubocop:disable RSpec/BeforeAfterAll, RSpec/InstanceVariable
+      before(:all) do
+        @original_dsi_env = ENV["DSI_ENV"]
```
Comment (peterdavidhamilton) [inline spec/lib/self-serve/dsi/uri_spec.rb:107]: Does this work? `ClimateControl.modify(DSI_ENV: nil) { example.run }`

Comment (peterdavidhamilton) [reply spec/lib/self-serve/dsi/uri_spec.rb:107]: Previous example used 
```ruby
around do |example|
  ClimateControl.modify(DSI_ENV: "test") { example.run }
end
```

File: spec/lib/self-serve/dsi/uri_spec.rb (hunk 3 above)
Comment (peterdavidhamilton) [reply spec/lib/self-serve/dsi/uri_spec.rb:6]: I'll do a version at a later date to see what you think

File: lib/dsi/client.rb [hunk 5]
```diff
@@ -2,17 +2,19 @@
 
 require "jwt"
 require "dry-initializer"
+
+require "types"
 require "dsi/user"
+# require "dsi/url"
```
Comment (peterdavidhamilton) [inline lib/dsi/client.rb:8]: `require "dsi/uri"`

File: lib/dsi/uri.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+require "types"
+
+# Generate Dfe Sign In URLs for the active environment:
+# production (default), pre-production, test
+#
+module Dsi
+  class Uri
+    extend Dry::Initializer
+
+    # @!attribute [r] env
```
Comment (peterdavidhamilton) [inline lib/dsi/uri.rb:12]: The `[r]` is the syntax for read-only so can be removed

**Review by peterdavidhamilton — CHANGES_REQUESTED** [review]
Can you address the comments please but leave the spec and I'll do a comparison

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -10,7 +10,7 @@ def banner_message
     I18n.t("banner.beta.message", support_email: ENV.fetch("SUPPORT_EMAIL"))
   end
 
-  def dsi_url(args = {})
-    ::Dsi::Uri.new(**args).call.to_s
+  def dsi_url(**args)
+    ::Dsi::Uri.new(args).call.to_s
```
Comment (peterdavidhamilton) [inline app/helpers/application_helper.rb:14]: Can we splat too?

Comment (d-a-v-e) [reply app/helpers/application_helper.rb:14]: does it make sense to splat it if it's already been splatted (this is a very normal sentence to write)

**Review by peterdavidhamilton — APPROVED** [review]

**Review by kennyevil — APPROVED** [review]

**Pull request merged**